src/main.rs

```diff
@@ -1,3 +1,207 @@
+    fn from(value: String) -> Self {
+        Token(value.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Chat<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    chat_type: ChatType,
+    id: u128,
+    messages: Vec<Message<'a>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum ChatType {
+    PublicChannel,
+    PrivateChannel,
+    PublicSupergroup,
+    PrivateSupergroup,
+    PersonalChat,
+    ChatForbidden,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Message<'a> {
+    id: u64,
+    #[serde(rename = "type")]
+    message_type: MessageType,
+    date: NaiveDateTime,
+    from: Option<Person<'a>>,
+    text_entities: Vec<TextEntity>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+enum MessageType {
+    Service,
+    Message,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TextEntity {
+    #[serde(rename = "type")]
+    text_type: TextEntityType,
+    text: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+enum TextEntityType {
+    Pre,
+    Bold,
+    Link,
+    Code,
+    Email,
+    Plain,
+    Phone,
+    Italic,
+    Cashtag,
+    Spoiler,
+    Mention,
+    Hashtag,
+    TextLink,
+    Underline,
+    BotCommand,
+    CustomEmoji,
+    MentionName,
+    Strikethrough,
+}
+
+impl TextEntityType {
+    pub fn is_meta(&self) -> bool {
+        use TextEntityType::*;
+        matches!(self, Phone | BotCommand | Email | CustomEmoji | Mention)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ChatStatistics<'a> {
+    num_tokens: usize,
+    #[serde(borrow)]
+    members_tokens_map: HashMap<Person<'a>, HashMap<Token<'a>, usize>>,
+    #[serde(borrow)]
+    tokens_map: HashMap<Token<'a>, usize>,
+}
+impl<'a> ChatStatistics<'a> {
+    pub fn gather(chat: &'a Chat) -> Self {
+        let jobs = current_num_threads();
+
+        let num_messages = chat.messages.len();
+        let messages_per_thread = num_messages / jobs;
+
+        let (sender, receiver) = crossbeam::channel::bounded(jobs + 1);
+        let mut tokens_map = HashMap::new();
+        let mut members_tokens_map = HashMap::new();
+
+        chat.messages.par_iter().chunks(messages_per_thread).for_each_with(sender, move |sender, messages: Vec<&Message>| {
+            let mut chunk_tokens_map: HashMap<Token<'_>, usize> = HashMap::new();
+            let mut chunk_members_tokens_map: HashMap<Person<'_>, HashMap<Token<'_>, usize>> = HashMap::new();
+            for message in messages.iter().filter(|message| message.message_type != MessageType::Service) {
+                let from = message.from.clone().unwrap();
+                for entity in message.text_entities.iter().filter(|entity| !entity.text_type.is_meta()) {
+                    for token in entity.text.split([' ', ',', '.','(', ')', '-', '!', '?', '\'', '\"', '\n', '\t']).filter(|s| !s.is_empty()) {
```

Comment:
Maybe it's reasonable to store that filter list in a `const`?

Reply:
That was my thought! 😅